vw units doesn't get updated inside shadow trees as frame is resized

rniwa opened the pull request:
#### 183e16cd8e477213dff4b6a30f78d95e21c528fa
<pre>
vw units doesn&apos;t get updated inside shadow trees as frame is resized
<a href="https://bugs.webkit.org/show_bug.cgi?id=216240">https://bugs.webkit.org/show_bug.cgi?id=216240</a>

Reviewed by Simon Fraser.

The bug was caused by Document::updateViewportUnitsOnResize only updating the style resolver of
the document and invalidating styles of nodes in the document tree. Fixed the bug by also
updating the style resolver and invalidating styles of nodes in each shadow tree.

* Source/WebCore/dom/Document.cpp:
(WebCore::Document::updateViewportUnitsOnResize):
* Source/WebCore/style/StyleScope.cpp:
(WebCore::Style::Scope::didChangeViewportSize): Added. Moved from updateViewportUnitsOnResize.
* Source/WebCore/style/StyleScope.h:

* LayoutTests/fast/shadow-dom/shadow-style-invalidation-vw-units-expected.html: Added.
* LayoutTests/fast/shadow-dom/shadow-style-invalidation-vw-units.html: Added.

Canonical link: <a href="https://commits.webkit.org/252148@main">https://commits.webkit.org/252148@main</a>
</pre>

smfr commented on LayoutTests/fast/shadow-dom/shadow-style-invalidation-vw-units.html:
```diff
@@ -0,0 +1,30 @@
+<!DOCTYPE html>
```
Could this be a Web Platform Test?

rniwa replied:
Sure, with a bit of modification to get rid of testRunner.waitUntilDone / notifyDone.

rniwa replied:
```html
<!DOCTYPE html>
<html>
<body>
<script>

const iframe = document.createElement('iframe');
iframe.scrolling = 'no';
iframe.style = 'border: none; width: 50px; height: 50px; overflow: hidden';
iframe.src = `data:text/html,<!DOCTYPE html><style>head,html,body { padding: 0; margin: 0; }</style>
<div style="width: 100vw; height: 50vw; background: green"></div>
<div style="width: 100px; height: 100px; background: red;"><div id="host"></div></div>
<script>host.attachShadow({mode: 'closed'}).innerHTML =
    '<style> div { width: 100vw; height: 50vw; background: green; }</style><div></div>';
host.getBoundingClientRect();
</sc` + `ript>`;
iframe.onload = () => {
    iframe.style.width = '100px';
    iframe.style.height = '100px';
}
document.body.appendChild(iframe);

</script>
</body>
</html>
```

rniwa committed:
vw units doesn't get updated inside shadow trees as frame is resized
https://bugs.webkit.org/show_bug.cgi?id=216240

Reviewed by Simon Fraser.

The bug was caused by Document::updateViewportUnitsOnResize only updating the style resolver of
the document and invalidating styles of nodes in the document tree. Fixed the bug by also
updating the style resolver and invalidating styles of nodes in each shadow tree.

* Source/WebCore/dom/Document.cpp:
(WebCore::Document::updateViewportUnitsOnResize):
* Source/WebCore/style/StyleScope.cpp:
(WebCore::Style::Scope::didChangeViewportSize): Added. Moved from updateViewportUnitsOnResize.
* Source/WebCore/style/StyleScope.h:

* LayoutTests/fast/shadow-dom/shadow-style-invalidation-vw-units-expected.html: Added.
* LayoutTests/fast/shadow-dom/shadow-style-invalidation-vw-units.html: Added.

Canonical link: https://commits.webkit.org/252148@main
